cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1127,6 +1127,42 @@ class ZeroCopyCast : public CastKernelBase {
   }
 };
 
+class ExtensionCastKernel : public CastKernelBase {
+ public:
+  static Status Make(const DataType& in_type, std::shared_ptr<DataType> out_type,
+                     const CastOptions& options,
+                     std::unique_ptr<CastKernelBase>* kernel) {
+    if (in_type.id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
+    auto storage_type = checked_cast<const ExtensionType&>(in_type).storage_type();
+
+    std::unique_ptr<UnaryKernel> storage_caster;
+    RETURN_NOT_OK(GetCastFunction(*storage_type, out_type, options, &storage_caster));
+    kernel->reset(
+        new ExtensionCastKernel(std::move(storage_caster), std::move(out_type)));
+
+    return Status::OK();
+  }
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(input.kind(), Datum::ARRAY);
+    if (input.type()->id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
```

Comment:
This should probably also verify that the `storage_type` (and maybe the `extension_name`) is the same as the input type the kernel was constructed with.

Reply:
Updated.